sdk/src/Services/DynamoDBv2/Custom/DataModel/_async/IDynamoDBContext.Async.cs

```diff
@@ -308,17 +387,54 @@ partial interface IDynamoDBContext
+        /// <param name="values">
+        /// Value(s) of the condition.
+        /// For all operations except QueryOperator.Between, values should be one value.
+        /// For QueryOperator.Betwee, values should be two values.
+        /// </param>
+        /// <param name="queryConfig">Config object that can be used to override properties on the table's context for this request.</param>
+        /// <returns>AsyncSearch which can be used to retrieve DynamoDB data.</returns>
+        AsyncSearch<T> QueryAsync<T>(object hashKeyValue, QueryOperator op, IEnumerable<object> values, QueryConfig queryConfig);
+
+        /// <summary>
+        /// Configures an async Query operation against DynamoDB using a mid-level document model 
+        /// query configration, finding items that match the specified conditions. 
+        /// </summary>
+        /// <typeparam name="T">Type of object.</typeparam>
+        /// <param name="queryConfig">Mid-level, document model query request object.</param>
+        /// <returns>AsyncSearch which can be used to retrieve DynamoDB data.</returns>
+        AsyncSearch<T> FromQueryAsync<T>(QueryOperationConfig queryConfig);
 
         /// <summary>
         /// Configures an async Query operation against DynamoDB, finding items
         /// that match the specified conditions.
         /// </summary>
         /// <typeparam name="T">Type of object.</typeparam>
-        /// <param name="queryConfig">Query request object.</param>
+        /// <param name="queryOperationConfig">Mid-level, document model query request object.</param>
         /// <param name="operationConfig">Config object which can be used to override the table used.</param>
         /// <returns>AsyncSearch which can be used to retrieve DynamoDB data.</returns>
-        AsyncSearch<T> FromQueryAsync<T>(QueryOperationConfig queryConfig, DynamoDBOperationConfig operationConfig = null);
+        [Obsolete("Use the FromQueryAsync overload that takes QueryConfig instead, since DynamoDBOperationConfig contains properties that are not applicable to FromQueryAsync.")]
+        AsyncSearch<T> FromQueryAsync<T>(QueryOperationConfig queryOperationConfig, DynamoDBOperationConfig operationConfig);
+
+        /// <summary>
+        /// Configures an async Query operation against DynamoDB using a mid-level document model 
+        /// query configration, finding items that match the specified conditions. 
+        /// </summary>
+        /// <typeparam name="T">Type of object.</typeparam>
+        /// <param name="queryOperationConfig">Mid-level, document model query request object.</param>
+        /// <param name="queryConfig">Config object that can be used to override properties on the table's context for this request.</param>
+        /// <returns>AsyncSearch which can be used to retrieve DynamoDB data.</returns>
+        AsyncSearch<T> FromQueryAsync<T>(QueryOperationConfig queryOperationConfig, QueryConfig queryConfig);
```

Comment:
This one is a little awkward because there are two similar query config objects.

1. We have an API that accepted the _document_ `QueryOperationConfig` object, even though we're in the object persistence model.
2. I didn't want to add everything that is now in `QueryConfig` to `QueryOperationConfig`, since `QueryConfig` is still used by the `Document` namespace and some settings won't be relevant.

I'm not sure about the motivation for this. We could explore ensuring the new `QueryConfig` has everything on `QueryOperationConfig` (right now it doesn't), then obsolete this one as well?

Reply:
From internal discussion on 7/11, leaning towards obsoleting this one as well, and driving towards a `FromQueryAsync<T>(QueryConfig queryConfig);`

Reply:
Looked at this again this week, I'm still struggling to collapse this down to a single input operation. 
* I don't want to expand the existing `QueryOperationConfig`/`ScanOperationConfig` with properties from `DynamoDBOperationConfig`, because some of those properties are only relevant to the object-persistence model.
* We previously added an explicit `FromQuery` which takes the _document_ `QueryOperationConfig` (and the scan counterpart). I'm not sure if this was an explicit ask, or we were reusing code, but I still think it may be valid to have a query or scan representation which can be used in both programming models. 
    * I don't want to take that capability away by deprecating or switching `FromQuery` to a new object.
* We could take all of the query/scan properties from `QueryOperationConfig`/`ScanOperationConfig` and move them to the new `QueryConfig`/`ScanConfig`.
    * But that defeats the purpose of `FromQuery`, since then it won't be much different from the new `Query(QueryConfig queryConfig)`

Reply:
New attempt via https://github.com/aws/aws-sdk-net/pull/3383/commits/987fe702d9428657cff4228bf91457a1ae9dcd74

The `From...` variants will now take:
1. The existing, document-model query or scan input, which controls the majority of the query/scan.
2. A new `FromQueryConfig` or `FromScanConfig`, which controls the table-level properties and object-persistence serialization.

This allow us:
* Continue to reuse the "main" query/scan inputs for both programming models
* Avoids any overlap between `QueryOperationConfig` (existing) and the new `FromQueryConfig`